pkg/scheduler/framework/plugins/podtopologyspread/scoring.go

```diff
@@ -175,7 +175,14 @@ func (pl *PodTopologySpread) PreScore(
 			if tpCount == nil {
 				continue
 			}
-			count := countPodsMatchSelector(nodeInfo.Pods, c.Selector, pod.Namespace)
+			count := 0
+			if len(pod.OwnerReferences) > 0 {
+				if v, ok := nodeInfo.PodsWithOwnerRef[pod.OwnerReferences[0].UID]; ok {
+					count = len(v)
```

Comment:
This assumption is invalid. `countPodsMatchSelector` returns number of pods on the node that match the namespace and selector. It's not equal to number of pods with the same owner reference. I think you would need to modify the `countPodsMatchSelector` to cache the `selector.Matches(labels.Set(p.Pod.Labels))` for given OwnerReference id, but now I'm not sure if it will give us a noticeable speedup.

Reply:
When we use deployment with topologySpreadConstraint, in order to spread the workload within the service.

The label selector specified in topologySpreadConstraint are closely matched with the selector in deployment spec.

This is the same logic applied to users of daemonset/statefulset, or other kind of workload controller.

In these ways, this assumption are always valid: 
only those pod owned by the same workload controller need to be accounted  in topology spreading

Could you provide me an example that, a pod matches these criterias:
1.  has an owner or to say, a controller
2.  with topologySpreadConstraint specified
But the label selector in topologySpreadConstraint, is not the same as selector specified in the controller's spec?

If this kind of case exists, I would also agree my assumption is not correct.

Yes you're right, make a cache of  selector.Matches(labels.Set(p.Pod.Labels)) indexed by owner id, will not have any speedup, I've also try that.

Reply:
> In the user scenario of deployment with topologySpreadConstraint,
> 
> the label selector specified in topologySpreadConstraint,
> 
> are closely matched with the distinguished set of selector in deployment spec.
> 
> So this assumption is always valid in below close-looped logic:
> 
> * only those pods owned by the same deployment, need to be accounted  in topology spreading
> * those pods required to be accounted in topology spreading, are owned by the same deployment
> 
> They also apply to any other scenarios in practical, which match below criterias:
> 
> * pods has topologySpreadConstraint specified
> * they are are also managed/owned by a workload controller (daemonset/statefulset/cloneset etc)
> 
> If an practical example can be provided that:
> 
> 1. a pod has an owner and topologySpreadConstraint
> 2. the label selector of topologySpreadConstraint is not the distinguished set of owner's selector.
> 
> I would also agree that my assumption is incorrect.
> 
> Yes you're right, making a cache of selector.Matches(labels.Set(p.Pod.Labels)) indexed by owner id, will not have any speedup, I've also tried that.

@sanposhiho could you also comment on this?

Reply:
@sanposhiho @denkensk could you also comment on this?

Reply:
> the label selector specified in topologySpreadConstraint,
are closely matched with the distinguished set of selector in deployment spec.

> So this assumption is always valid in below close-looped logic:
- only those pods owned by the same deployment, need to be accounted in topology spreading
- those pods required to be accounted in topology spreading, are owned by the same

Those are not always correct; multiple deployments might share the same topology spreading.
And, we cannot implement something that could work wrong, even in a minor scenario.

Reply:
@sanposhiho  I can see your point of view. 

But in practical usage of topology spreading, we never find a rollout pattern, that shares a same topologySpreadConstaint across different deployments/statefulset/cloneset.

If that happens, it means there's a chance that, topologySpreadConstaint can be abstracted outside the scope of pod spec,  in order to be reused more easily.

But I can see that happen in a more-acceptable way, since topologySpreadConstaint are always
used within the same scope of pods, likewise, pods with a same controller.

Perhaps this could be a cross-boundary topic between sig-scheduling and sig-apps, which may never be settled.

Reply:
If youre suggesting to change the topology spread like that, that's a huge breaking change and cannot be acceptable.

Reply:
OK, that's understandable.

Reply:
OK, that's understandable, in your point of view.

Reply:
I've applied this in our production clusters, it works like a charm.

However, I have full control over the service rollout on these cluster, so I think the gab between us lies in this difference.

Thanks so much for your time.

Reply:
Can you close the PR if you agree it cannot be merged due to the breaking change concern? (Or course, you can still seek other opportunities in the topology spreading to improve the performance, without breaking)

Reply:
I want it to be spread out to more people who actually need it. But of course, I'll close it finally.